docs/docs/typelevel.md

```diff
@@ -0,0 +1,449 @@
+  case _: Unit => Some(())
+  case _: t >: Null => Some(null)
+  case _ => None
+}
+```
+Then:
+```scala
+defaultValue[Int] = Some(0)
+defaultValue[Boolean] = Some(false)
+defaultValue[String | Null] = Some(null)
+defaultValue[AnyVal] = None
+```
+
+As another example, consider the type-level inverse of `toNat`: given a _type_ representing a Peano number, return the integer _value_ corresponding to it. Here's how this can be defined:
+```scala
+transparent def toInt[N <: Nat]: Int = anyValue[N] match {
+  case _: Z => 0
+  case _: S[n] => toInt[n] + 1
+}
+```
+
+## Computing New Types
+
+The examples so far all computed _terms_ that have interesting new types. Since in Scala terms can contain types it is equally easy to compute the types of these terms directly as well. To this purpose, it is helpful
+to base oneself on the helper class `scala.typelevel.Typed`, defined as follows:
+```scala
+class Typed[T](val value: T) { type Type = T }
+```
+Here is a version of `concat` that computes at the same time a result and its type:
+```scala
+transparent def concatTyped(xs: Tuple, ys: Tuple): Typed[_ <: Tuple] = xs match {
+  case ()       => Typed(ys)
+  case (x, xs1) => Typed((x, concatTyped(xs1, ys).value))
+}
+```
+
+## Avoiding Code Explosion
+
+Recursive transparent functions implement a form of loop unrolling through inlining. This can lead to very large generated expressions. The code explosion can be avoided by calling typed versions of the transparent functions to define erased values, of which  just the typed part is used afterwards. Here is how this can be done for `concat`:
+
+```scala
+def concatImpl(xs: Tuple, ys: Tuple): Tuple = xs match {
+  case ()       => ys
+  case (x, xs1) => (x, concatImpl(xs1, ys))
+}
+
+transparent def concat(xs: Tuple, ys: Tuple): Tuple = {
+  erased val r = concatTyped(xs, ys)
+  concatImpl(xs, ys).asInstanceOf[r.Type]
+}
+```
+The transparent `concat` method makes use of two helper functions, `concatTyped` (described in the last section) and `concatImpl`. `concatTyped` is called as the right hand side of an `erased` value `r`. Since `r` is `erased`, no code is generated for its definition.
+`concatImpl` is a regular, non-transparent function that implements `concat` on generic tuples. It is not inlineable, and its result type is always `Tuple`. The actual code for `concat` calls `concatImpl` and casts its result to type `r.Type`, the computed result type of the concatenation. This gives the best of both worlds: Compact code and expressive types.
+
+One might criticize that this scheme involves code duplication. In the example above, the recursive `concat` algorithm had to be implemented twice, once as a regular function, the other time as a transparent function. However, in practice it is is quite likely that the regular function would use optimized data representatons and algortihms that do not lend themselves easily to a typelevel interpretation. In these cases a dual implementation is required anyway.
+
```

Comment:
Note that the need for this duplicative and unsafe-feeling code could be avoided if we kept `inline` around and `transparent` _didn't_ imply `inline`: that is, if `transparent` by itself caused the code to get run both at compile time *and* at runtime, whereas `inline transparent` caused it to work exactly as this proposal currently describes. Of course, we'd probably have to either rule out implicit matches or have them only consider the definition site (and thus be kind of useless) without `inline`...

Reply:
How do you do term reduction (necessary for the resulting type refinement) without at least inlining the argument into the body?
This is why I was thinking of inlining the argument into the body, looking at what's left from it at after the reductions and using that as parameters to a specialized version of the transparent function, specialized version which would be cached and reused by invocations that reduce similarly. But [caching may be too tricky](https://github.com/lampepfl/dotty/pull/4616#issuecomment-402982343).

Reply:
Hmm, yes yes! If I understand, you make a good point: you could quickly get an explosion of runtime versions of the method! Caching seems like it could help but it would only reduce the problem rather than eliminate it... It might also be really hard to get binary compatibility right... It's almost like this wouldn't be practical unless it were paired with the behavior of _not_ recomputing invocation receivers at the use site that I described in my other comment... slash... Am I somehow missing a way that you would still somehow get runtime-version explosion if it worked that way?

Reply:
> How do you do term reduction (necessary for the resulting type refinement) without at least inlining the argument into the body?

@LPTK By working on types directly instead of manipulating untyped trees, which is what the alternative "TypeOf" thing is about. Let me illustrate with an example.

```scala
transparent def unroll(i: Int): String = {
  println(i)
  if (i == 0) "foo"
  else unroll(i - 1)
}
```

With transparent implemented with inlining of untyped trees, `unroll(2)` would be expanded as follows:

```scala
unroll(2)

--> (inlining)
{
  val i = 2
  println(i)
  if (i == 0) "foo"
  else unroll(i - 1)
}

--> (constant foldings)
{ println(2); unroll(1) }

--> (inlining)
{ println(2); {
  val i = 1
  println(i)
  if (i == 0) "foo"
  else unroll(i - 1)
}}

--> (constant foldings)
{ println(2); println(1); unroll(0) }

--> (inlining)
{ println(2); println(1); {
  val i = 0
  println(i)
  if (i == 0) "foo"
  else unroll(i - 1)
}}

--> (constant foldings)
{ println(2); println(1); println(0); "foo" }
```

The resulting block is then re-typechecked to finally compute the very precise singleton type "foo" for `unroll(2)`.

Now in the typed approach, the definition of unroll is typechecked once where it's defined, and terms are never touched again. Here is how the definition of unroll would be typed, working bottom up:

```scala
transparent def unroll(i: Int): String = {
  println(i) // : Unit
  if (
    i == 0   // : { i.type == 0 }
  )
    "foo"    // : "foo"
  else
    unroll(
      j      // : { i.type - 1 }
    )        // : { unroll({ i.type - 1 }) }
}
```

Where `{ a.f(b) }` is special syntax for the precise type of the call to `f`. Don't be confused by the presence of `{.}`, this is really a type (composed of other types), they are not trees in there. A more verbose syntax would be `CALL-TYPE(a.type, f.symbol, b.type)`.

The `if` would then be typed as follows:

```scala
IF-TYPE({ i.type == 0 }, "foo", { unroll({ i.type - 1 }) })
```

The block would get the same type, which thus becomes the precise return type of `unroll`. The call to `unroll(2)` would then be types as `{ unroll(2) }`, which can be normalized as follows:

```scala
{ unroll(2) }

(β-reduction) -->
IF-TYPE({ i.type == 0 }, "foo", { unroll({ i.type - 1 }) }) [2/i.type]

(substitution) -->
IF-TYPE({ 2 == 0 }, "foo", { unroll({ 2 - 1 }) })

(constant folding) -->
IF-TYPE(false, "foo", { unroll(1) })

(if) -->
{ unroll(1) }

...

--> "foo"
```

In conclusion: inlining is not mandatory to do computations on types as described in the proposal. Of course working with types is less powerful as it doesn't do any code specialization. The `TypeOf` approach is, in a sense, simpler and less ambitious.